## DATA Cleaning and EDA

In [1]:
# Import the libraries we will need
import pandas as pd
import numpy as np
import scipy.stats as stats
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib.pyplot as plt
import math

# setting the visual
sns.set(font_scale=2)
cmap = sns.diverging_palette(220, 10, as_cmap=True) # one of the many color mappings
sns.set_style('whitegrid')
%config InlineBackend.figure_format = 'retina'
%matplotlib inline  

In [2]:
df = pd.read_csv('./input/dcrime.csv')
ward_data =pd.read_csv('./input/warddata.csv')

In [3]:
print (df.duplicated().sum())

0


# Checking Data Integrity

In [4]:
print(df.shape)
print(df.head(2))
print(ward_data.shape)
print(ward_data.head(2))

(231502, 25)
   Unnamed: 0          X          Y       CCN                REPORT_DAT  \
0           0 -76.999518  38.901924   9074624  2012-04-25T00:00:00.000Z   
1           1 -76.995541  38.905032  10123633  2012-02-29T00:00:00.000Z   

      SHIFT  METHOD    OFFENSE                                 BLOCK  \
0  MIDNIGHT  OTHERS  SEX ABUSE      900 - 999 BLOCK OF 5TH STREET NE   
1  MIDNIGHT  OTHERS  SEX ABUSE  700 - 799 BLOCK OF FLORIDA AVENUE NE   

     XBLOCK    ...      NEIGHBORHOOD_CLUSTER  BLOCK_GROUP CENSUS_TRACT  \
0  400042.0    ...                Cluster 25     010600 2      10600.0   
1  400387.0    ...                Cluster 25     010600 1      10600.0   

   VOTING_PRECINCT   LATITUDE  LONGITUDE  BID                START_DATE  \
0      Precinct 83  38.901916 -76.999516  NaN  2009-05-31T23:00:00.000Z   
1      Precinct 83  38.905024 -76.995538  NaN  2010-08-27T02:30:00.000Z   

                   END_DATE   OBJECTID  
0  2009-06-01T06:00:00.000Z  167253019  
1  2010-08-27

In [5]:
df.isnull().sum().sort_values(ascending=False)  

BID                     192953
END_DATE                  5647
NEIGHBORHOOD_CLUSTER      2843
BLOCK_GROUP                676
CENSUS_TRACT               676
PSA                        135
DISTRICT                   129
VOTING_PRECINCT             47
WARD                         8
START_DATE                   4
REPORT_DAT                   0
CCN                          0
OBJECTID                     0
METHOD                       0
Y                            0
X                            0
SHIFT                        0
ANC                          0
OFFENSE                      0
BLOCK                        0
XBLOCK                       0
YBLOCK                       0
LATITUDE                     0
LONGITUDE                    0
Unnamed: 0                   0
dtype: int64

In [6]:
# checking the null values of the df and sorting them as percentage of the DF shape
(df.isnull().sum()/df.shape[0]).sort_values(ascending=False)  

BID                     0.833483
END_DATE                0.024393
NEIGHBORHOOD_CLUSTER    0.012281
BLOCK_GROUP             0.002920
CENSUS_TRACT            0.002920
PSA                     0.000583
DISTRICT                0.000557
VOTING_PRECINCT         0.000203
WARD                    0.000035
START_DATE              0.000017
REPORT_DAT              0.000000
CCN                     0.000000
OBJECTID                0.000000
METHOD                  0.000000
Y                       0.000000
X                       0.000000
SHIFT                   0.000000
ANC                     0.000000
OFFENSE                 0.000000
BLOCK                   0.000000
XBLOCK                  0.000000
YBLOCK                  0.000000
LATITUDE                0.000000
LONGITUDE               0.000000
Unnamed: 0              0.000000
dtype: float64

In [7]:
ward_data.isnull().sum()/len(ward_data)

year              0.0
ANC               0.0
Housing_Prices    0.0
Unemployment      0.0
dtype: float64

####  Feature Summary Statistics

###### Count of crimes committed per duty shift
The Evening shift had the highest count as compared to the three categories followed by the Day shift. However, this doesn't take into account the severity of the crime.

In [8]:
print('==================================================')
print( "---===<<< Duty Shift: variable = 'SHIFT' >>>===--")
var_count = df.groupby('SHIFT')
print( var_count.CCN.count())
print('================================================')

---===<<< Duty Shift: variable = 'SHIFT' >>>===--
SHIFT
DAY         88724
EVENING     98882
MIDNIGHT    43896
Name: CCN, dtype: int64


###### Count of crimes committed per Police District
The number of crimes is evenly distributed across all the PSA areas ignoring the severity of the crime.

In [9]:
print('=========================================================================')
print("---==< Crimes committed per police district: variable = 'DISTRICT' >==---")
var_count = df.groupby('DISTRICT')
print(var_count.CCN.count())
print('=========================================================================')

---==< Crimes committed per police district: variable = 'DISTRICT' >==---
DISTRICT
1.0    39865
2.0    35581
3.0    39055
4.0    30492
5.0    32821
6.0    29807
7.0    23752
Name: CCN, dtype: int64


Below is a sample output from a single record to get a feel for the type of data.

In [10]:
print('=========================================================================')
print("---==< Crimes|OFFENSES committed per BLOCK: variable = 'BLOCK' >==---")
print(df.groupby('BLOCK').OFFENSE.value_counts())
print('=========================================================================')

---==< Crimes|OFFENSES committed per BLOCK: variable = 'BLOCK' >==---
BLOCK                                          OFFENSE                   
0 - 0 BLOCK OF 14TH STREET SW                  THEFT/OTHER                    1
0 - 0 BLOCK OF ARLINGTON MEMORIAL BRIDGE SW    THEFT F/AUTO                   1
                                               THEFT/OTHER                    1
0 - 0 BLOCK OF CHEVY CHASE CIRCLE NW           BURGLARY                       3
                                               THEFT F/AUTO                   2
                                               ROBBERY                        1
                                               THEFT/OTHER                    1
0 - 0 BLOCK OF COLUMBUS CIRCLE NE              THEFT/OTHER                   11
                                               ASSAULT W/DANGEROUS WEAPON     1
                                               ROBBERY                        1
0 - 0 BLOCK OF HEALY CIRCLE NW                 THEFT/OTH

In [11]:
print('======================================================')
print('--====< Example Record from Data Set>============')
print(df.iloc[1234])
print('======================================================')

--====< Example Record from Data Set>============
Unnamed: 0                                        1234
X                                             -77.0028
Y                                              38.9141
CCN                                           12008660
REPORT_DAT                    2012-01-18T09:32:00.000Z
SHIFT                                              DAY
METHOD                                          OTHERS
OFFENSE                            MOTOR VEHICLE THEFT
BLOCK                   200 - 299 BLOCK OF S STREET NE
XBLOCK                                          399759
YBLOCK                                          138467
WARD                                                 5
ANC                                                 5E
DISTRICT                                             5
PSA                                                502
NEIGHBORHOOD_CLUSTER                        Cluster 21
BLOCK_GROUP                                   008702 2
CENSUS_TRACT   

##### Verify data quality

##### Unique Data

In [12]:
print('======================================================')
print (np.count_nonzero(df['BLOCK'].unique()))
df.drop('BLOCK', axis=1, inplace=True)

print (np.count_nonzero(df['BLOCK_GROUP'].unique()))
df.drop('BLOCK_GROUP', axis=1, inplace=True)
print('======================================================')

14245
451


Block has 14,245 unique values indicating the street and block identifier. This value appears to be based on the CENSUS_TRACT variable, but with higher resolution. Mapping these unique values is time consuming and doesn't provide meaningful value for for further Analysis. Therefore, the column will be removed from the working data set.
Block Group has 451 unique values indicating the land management tract identifier. Mapping these unique values is time consuming and doesn't provide meaning value for further Analysis. Therefore, the column will be removed from the working data set.

In [13]:
# if END_DATE is NaN, then use START_DATE
df['END_DATE'].fillna(df['START_DATE'], inplace=True)

# if VOTING_PRECINCT is NaN, then set it to 0
df['VOTING_PRECINCT'].fillna(0, inplace=True)

# if NEIGHBORHOOD_CLUSTER is NaN, then set it to 0
df['NEIGHBORHOOD_CLUSTER'].fillna(0, inplace=True)

# if CENSUS_TRACT is NaN, then set it to 0
df['CENSUS_TRACT'].fillna(0, inplace=True)
# if WARD is NaN, then use ANC
# df['WARD'].fillna(df['ANC'], inplace=True)
# if WARD is NaN, then set it to 0
df['WARD'].fillna(0, inplace=True)

In [14]:
#  Create a dataframe that holds the central location of each Police Service Area (PSA)
#  The PSAs: according to the information from the DC metropolitan ionformation and opendata DC
# PSA's are smaller than the Police Districts, so we should have better accuracy in identifying the associated PSA
#  The PSA ID contains the District ID, so we can impute the District from the PSA
psa_loc = pd.DataFrame(df[['PSA','XBLOCK','YBLOCK']].groupby('PSA').median())

#  ---==< Estimate PSA membership based on proximity to each area's centroid >==---
def nearbyPSA(nPSA,dX,dY):
    # Default to the current PSA ID
    nearbyPSA = nPSA
    
    # Only operate on missing IDs
    if (pd.isnull(nPSA)):
        minDist = 9e99  # Set the initial closest distance to be a large value
        nearbyPSA = 0
        
        # Loop through the records in the psa_loc dataframe
        for PSA_ID, PSA in psa_loc.iterrows():
            # Calculate the distance between the report and the current PSA using the Eucleadian distance
            thisDist = math.sqrt((dX - PSA['XBLOCK'])**2 + (dY - PSA['YBLOCK'])**2)
            
            # If this distance is smaller than the current minimum distance, update the minimum distance
            if (thisDist < minDist):
                minDist = thisDist # Replace the minimum distance with the current distance
                nearbyPSA = PSA_ID # Remember which PSA this is related to
                
    # Return the ID for the closest PSA
    return [nearbyPSA, int(nearbyPSA / 100)]

#  Impute the missing PSA
df['PSA_ID'] = 0
df['DistrictID'] = 0
df[['PSA_ID','DistrictID']] = list(map(nearbyPSA,df['PSA'],df['XBLOCK'],df['YBLOCK']))

#  Display the results of this imputation method.
print('==========================================================')
print(df[['PSA','DISTRICT','PSA_ID','DistrictID']][df['PSA'].isnull()])
print('==========================================================')

        PSA  DISTRICT  PSA_ID  DistrictID
4575    NaN       NaN   708.0           7
9560    NaN       1.0   107.0           1
11114   NaN       NaN   208.0           2
11333   NaN       NaN   102.0           1
12890   NaN       NaN   506.0           5
14021   NaN       NaN   102.0           1
20148   NaN       NaN   704.0           7
22443   NaN       NaN   102.0           1
26647   NaN       NaN   202.0           2
26848   NaN       NaN   102.0           1
28337   NaN       NaN   102.0           1
29280   NaN       1.0   108.0           1
29566   NaN       1.0   108.0           1
29979   NaN       1.0   103.0           1
32311   NaN       NaN   102.0           1
33553   NaN       NaN   403.0           4
39115   NaN       NaN   102.0           1
46313   NaN       NaN   102.0           1
46530   NaN       NaN   102.0           1
46860   NaN       NaN   102.0           1
50480   NaN       NaN   103.0           1
54194   NaN       NaN   105.0           1
55129   NaN       NaN   102.0     

##### Checking for duplicated  data

In [15]:
# do we have duplicate rows?
# the great thing is that i took care of them while concatenating the different values in the initial Data setup
print(df.duplicated().sum())

0


In [16]:
df.WARD.isnull().sum()

0

#### Feature Reeengineering 

In [17]:
# strip 'Precinct ' from VOTING_PRECINCT values
# http://stackoverflow.com/questions/13682044/pandas-dataframe-remove-unwanted-parts-from-strings-in-a-column
df['VOTING_PRECINCT'] = df['VOTING_PRECINCT'].apply(str).map(lambda x: x.lstrip('Precinct '))

# strip 'Cluster ' from NEIGHBORHOOD_CLUSTER values
df['NEIGHBORHOOD_CLUSTER'] = df['NEIGHBORHOOD_CLUSTER'].apply(str).map(lambda x: x.lstrip('Cluster '))

In [18]:
print('======================================================')
# convert REPORT_DAT to datetime
df['REPORT_DAT'] = pd.to_datetime(df['REPORT_DAT'])

# convert SHIFT to int
shift_mapping = {'day':1, 'evening':2, 'midnight':3}
df['SHIFT_Code'] = df['SHIFT'].str.lower().map(shift_mapping).astype('category')

# convert OFFENSE to numeric
# Python for Data Analysis, pg. 279
offense_mapping = {'theft/other':1, 'theft f/auto':2, 'burglary':3, 'assault w/dangerous weapon':4, 'robbery':5,
                  'motor vehicle theft':6, 'homicide':7, 'sex abuse':8, 'arson':9}
df['OFFENSE_Code'] = df['OFFENSE'].str.lower().map(offense_mapping).astype('category')
df['OFFENSE'] = df['OFFENSE'].str.replace('DANGEROUS WEAPON', 'DW')

# convert METHOD to numeric
method_mapping = {'others':1, 'gun':2, 'knife':3}
df['METHOD_Code'] = df['METHOD'].str.lower().map(method_mapping).astype('category')

# convert DISTRICT to numeric
df['DistrictID'] = df['DistrictID'].astype(np.int64)

# convert PSA to numeric
df['PSA_ID'] = df['PSA_ID'].astype(np.int64)

# convert WARD to numeric
df['WARD'] = df['WARD'].astype(np.int64)

# convert ANC to numeric
anc_mapping = {'1B':12, '1D':14, '1A':11, '1C':13, '6E':65, '4C':43, '5E':55, '2B':22, '2D':24, '2F':26, '2C':23,
       '2E':25, '2A':21, '3C':33, '3E':35, '3B':32, '3D':34, '3F':36, '3G':37, '4A':41, '4B':42, '4D':44,
       '5A':51, '5D':54, '5C':53, '5B':52, '6A':61, '6C':63, '6B':62, '6D':64, '7D':74, '7C':73, '7E':75,
       '7B':72, '7F':76, '8A':81, '8B':82, '8C':83, '8D':84, '8E':85}
df['ANC'] = df['ANC'].map(anc_mapping).astype('category')

# convert NEIGHBORHOOD_CLUSTER to numeric
df['NEIGHBORHOOD_CLUSTER'] = df['NEIGHBORHOOD_CLUSTER'].astype(np.int64)

# convert CENSUS_TRACT to numeric
df['CENSUS_TRACT'] = df['CENSUS_TRACT'].astype(np.int64)

# convert VOTING_PRECINCT to numeric
df['VOTING_PRECINCT'] = df['VOTING_PRECINCT'].astype(np.int64)

# convert CCN to numeric
df['CCN'] = df['CCN'].astype(np.int64)

# convert XBLOCK, YBLOCK to numeric
df['XBLOCK'] = df['XBLOCK'].astype(np.float64)
df['YBLOCK'] = df['YBLOCK'].astype(np.float64)

# convert START_DATE, END_DATE to dateime
df['START_DATE'] = pd.to_datetime(df['START_DATE'])
df['END_DATE'] = pd.to_datetime(df['END_DATE'])
print('======================================================')
print(df.info())
print()
print(df.iloc[1234])
print('======================================================')


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231502 entries, 0 to 231501
Data columns (total 28 columns):
Unnamed: 0              231502 non-null int64
X                       231502 non-null float64
Y                       231502 non-null float64
CCN                     231502 non-null int64
REPORT_DAT              231502 non-null datetime64[ns]
SHIFT                   231502 non-null object
METHOD                  231502 non-null object
OFFENSE                 231502 non-null object
XBLOCK                  231502 non-null float64
YBLOCK                  231502 non-null float64
WARD                    231502 non-null int64
ANC                     231502 non-null category
DISTRICT                231373 non-null float64
PSA                     231367 non-null float64
NEIGHBORHOOD_CLUSTER    231502 non-null int64
CENSUS_TRACT            231502 non-null int64
VOTING_PRECINCT         231502 non-null int64
LATITUDE                231502 non-null float64
LONGITUDE               231502 n

##### How do we identify Outliers
one hundredsix crimes were reported in 2012 which started before 2011. Six of the crimes started in 1915. Are these start dates in error or were these unsolved crimes recently solved through new technology advancements? Is it truly a 100-year-old cold case, or did the person simply enter the wrong century? beacuse the timestamp shows from 01/16/1800 one crime, two crimes from 1912 reported in 2011 and one crime of 1914 reported in 2011.

In [19]:
print('======================================================')
print('---======< Dealing with outliers>=======-------')
print(df['START_DATE'][df['START_DATE']<'1/1/2011'].count())
print(sorted(df['START_DATE'][df['START_DATE']<'1/1/2011'])[:10])
print('======================================================')

---======< Dealing with outliers>=======-------
106
[Timestamp('1800-01-16 09:51:02'), Timestamp('1912-08-22 21:00:00'), Timestamp('1912-09-16 23:40:00'), Timestamp('1914-09-26 19:00:00'), Timestamp('1915-03-18 16:00:00'), Timestamp('1915-08-30 06:00:35'), Timestamp('1915-09-17 18:00:52'), Timestamp('1915-10-10 22:30:45'), Timestamp('1915-10-16 21:03:03'), Timestamp('1915-10-17 21:00:57')]


four hundred fourty Five crimes were reported in 2013 which started before 2012. Six of the crimes started in 1915. Are these start dates in error or were these unsolved crimes recently solved through new technology advancements? Is it truly a 100-year-old cold case, or did the person simply enter the wrong century? beacuse the timestamp shows from 01/16/1800 one crime, two crimes from 1912 reported in 2011 and one crime of 1914 reported in 2012.

In [20]:
print('======================================================')
print('---======< Dealing with outliers>=======-------')
print(df['START_DATE'][df['START_DATE']<'1/1/2012'].count())
print(sorted(df['START_DATE'][df['START_DATE']<'1/1/2012'])[:10])
print('======================================================')

---======< Dealing with outliers>=======-------
445
[Timestamp('1800-01-16 09:51:02'), Timestamp('1912-08-22 21:00:00'), Timestamp('1912-09-16 23:40:00'), Timestamp('1914-09-26 19:00:00'), Timestamp('1915-03-18 16:00:00'), Timestamp('1915-08-30 06:00:35'), Timestamp('1915-09-17 18:00:52'), Timestamp('1915-10-10 22:30:45'), Timestamp('1915-10-16 21:03:03'), Timestamp('1915-10-17 21:00:57')]


35633 crimes were reported in 2014 which started before 2013. Six of the crimes started in 1915. Are these start dates in error or were these unsolved crimes recently solved through new technology advancements? Is it truly a 100-year-old cold case, or did the person simply enter the wrong century? beacuse the timestamp shows from 01/16/1800 one crime, two crimes from 1912 reported in 2011 and one crime of 1914 reported in 2013.

In [32]:
print('======================================================')
print('---======< Dealing with outliers>=======-------')
print(df['START_DATE'][df['START_DATE']<'1/1/2013'].count())
print(sorted(df['START_DATE'][df['START_DATE']<'1/1/2013'])[:10])
print('======================================================')

---======< Dealing with outliers>=======-------
35633
[Timestamp('1800-01-16 09:51:02'), Timestamp('1912-08-22 21:00:00'), Timestamp('1912-09-16 23:40:00'), Timestamp('1914-09-26 19:00:00'), Timestamp('1915-03-18 16:00:00'), Timestamp('1915-08-30 06:00:35'), Timestamp('1915-09-17 18:00:52'), Timestamp('1915-10-10 22:30:45'), Timestamp('1915-10-16 21:03:03'), Timestamp('1915-10-17 21:00:57')]


###### Creating New Features
creating new feature CRIME_TYPE to indicate if the crime was violent or nonviolent. This feature will be used for later prediction purposes.

In [21]:
# creating feature for crime type 1 = Violent, 2 = non violent
violent_offense = [4, 5, 7, 8]
df['CRIME_TYPE'] = np.where(df['OFFENSE_Code'].isin(violent_offense), 1, 2)
df['CRIME_TYPE'] = df['CRIME_TYPE'].astype('category')

Creating additional feature AGE to indicate the timespan between the latest the crime could have been committed and earliest.

In [22]:
# create age of crime END_DATE - START_DATE in seconds
df['AGE'] = (df['END_DATE'] - df['START_DATE'])/np.timedelta64(1, 's')

In [23]:
# create new feature TIME_TO_REPORT to indicate the timespan between the latest time the crime was committed and
# the time it was reported to the MPD. that is  time it took from crime to report it, REPORT_DAT - END_DATE in seconds
df['TIME_TO_REPORT'] = (df['REPORT_DAT'] - df['END_DATE'])/np.timedelta64(1, 's')

In [24]:
df['DATE'] = pd.to_datetime(df['END_DATE'], format = '%d/%m/%Y %H:%M:%S')

In [25]:
def date_separate(df):
    """ function that separates the sepecific interest days,weeks from the END_DATE"""
    df = df.copy()
    df['Year'] = pd.DatetimeIndex(df['DATE']).year
    df['Month'] = pd.DatetimeIndex(df['DATE']).month
    df['Day'] = pd.DatetimeIndex(df['DATE']).day
    df['hour'] = pd.DatetimeIndex(df['DATE']).hour
    df['dayofyear'] =pd.DatetimeIndex(df['DATE']).dayofyear
    df['week'] =pd.DatetimeIndex(df['DATE']).week
    df['weekofyear'] =pd.DatetimeIndex(df['DATE']).weekofyear
    df['dayofweek'] =pd.DatetimeIndex(df['DATE']).dayofweek
    df['weekday'] =pd.DatetimeIndex(df['DATE']).weekday
    df['quarter'] =pd.DatetimeIndex(df['DATE']).quarter
    return df

In [26]:
# here i have created an additional columns that helps us in analyzing the timing of the crimes happeing either by day
# week, month, year as well as the quarters and weekdays and week ends.
df['Year'] = pd.DatetimeIndex(df['DATE']).year
df['Month'] = pd.DatetimeIndex(df['DATE']).month
df['Day'] = pd.DatetimeIndex(df['DATE']).day
df['hour'] = pd.DatetimeIndex(df['DATE']).hour
df['dayofyear'] =pd.DatetimeIndex(df['DATE']).dayofyear
df['week'] =pd.DatetimeIndex(df['DATE']).week
df['weekofyear'] =pd.DatetimeIndex(df['DATE']).weekofyear
df['dayofweek'] =pd.DatetimeIndex(df['DATE']).dayofweek
df['weekday'] =pd.DatetimeIndex(df['DATE']).weekday
df['quarter'] =pd.DatetimeIndex(df['DATE']).quarter

#### MOST IMPORTANT ATTRIBUTTES

###### OFFENSE

In [27]:
print('==============================================================')
print('---<Examine the frequency of types of crimes/offenses>--======')
print('    __________________________________________________')
print(' Total Offenses - Count')
total_crime = df.CCN.count()
print (total_crime)
print('')
crime_rate = df.groupby('OFFENSE')
print('_________________________________')
print (' Offense Type - Count')
print('_________________________________')
print (crime_rate.CCN.count())
print ('')
print('_________________________________')
print (' Offense as percentage of total ')
print ('_______________________________')
print (crime_rate.CCN.count() / total_crime * 100.0)
print()
print ('___________________________')
print (' Offense Rate per 100,000 ')
print ('__________________________')
print (crime_rate.CCN.count() / 6.93972 )
print()
print ('_____________________________________')
print (' Odds of being a victim - by offense')
print ('____________________________________')
print (crime_rate.CCN.count() / 693972)

print('==============================================================')

---<Examine the frequency of types of crimes/offenses>--======
    __________________________________________________
 Total Offenses - Count
231502

_________________________________
 Offense Type - Count
_________________________________
OFFENSE
ARSON                    127
ASSAULT W/DW           14528
BURGLARY               16965
HOMICIDE                 781
MOTOR VEHICLE THEFT    17714
ROBBERY                20868
SEX ABUSE               1989
THEFT F/AUTO           69604
THEFT/OTHER            88926
Name: CCN, dtype: int64

_________________________________
 Offense as percentage of total 
_______________________________
OFFENSE
ARSON                   0.054859
ASSAULT W/DW            6.275540
BURGLARY                7.328230
HOMICIDE                0.337362
MOTOR VEHICLE THEFT     7.651770
ROBBERY                 9.014177
SEX ABUSE               0.859172
THEFT F/AUTO           30.066263
THEFT/OTHER            38.412627
Name: CCN, dtype: float64

___________________________
 Offens

###### METHOD

In [28]:
print('==============================================================')
print('-<Examine the frequency of types of crimes/offenses by Method>===')
print ('_____________________________________')
print(' Total Methods - Count') 
print ('_____________________________________')
print(df.CCN.count())
print ('')

method_rate = df.groupby('METHOD')
print ('_____________________')
print ('Method Type - Count')
print ('____________________')
print(method_rate.CCN.count())
print()
print ('______________________________')
print (' Method as percentage of total ')
print ('_____________________________')
print (method_rate.CCN.count() / total_crime * 100.0)

print ('_________________________')
print (' Method Rate per 100,000 ')
print ('_________________________')
print( method_rate.CCN.count() / 6.93972)
print ('___________________________________')
print (' Odds of being a victim - by method')
print ('_______________________________')
print (method_rate.CCN.count() / 693972)
print('==============================================================')

-<Examine the frequency of types of crimes/offenses by Method>===
_____________________________________
 Total Methods - Count
_____________________________________
231502

_____________________
Method Type - Count
____________________
METHOD
GUN        13003
KNIFE       7234
OTHERS    211265
Name: CCN, dtype: int64

______________________________
 Method as percentage of total 
_____________________________
METHOD
GUN        5.616798
KNIFE      3.124811
OTHERS    91.258391
Name: CCN, dtype: float64
_________________________
 Method Rate per 100,000 
_________________________
METHOD
GUN        1873.706720
KNIFE      1042.405169
OTHERS    30442.870894
Name: CCN, dtype: float64
___________________________________
 Odds of being a victim - by method
_______________________________
METHOD
GUN       0.018737
KNIFE     0.010424
OTHERS    0.304429
Name: CCN, dtype: float64


##### Age of Crime


In [29]:
print('==============================================================')
print('-<Examine the age of  crimes/offenses by Time >===')

# hours
temp = df['AGE'] / 3600

print( temp.describe())
print('') 

# exclude temp two std away from mean
#http://stackoverflow.com/questions/23199796/detect-and-exclude-outliers-in-pandas-dataframe
print (temp[~(np.abs(temp - temp.mean())>(2*temp.std()))].describe())
print('==============================================================')


-<Examine the age of  crimes/offenses by Time >===
count    2.314980e+05
mean     1.017558e+02
std      7.671005e+03
min     -3.057575e+04
25%      8.333333e-02
50%      8.533333e-01
75%      7.000556e+00
max      1.875912e+06
Name: AGE, dtype: float64

count    231402.000000
mean         32.664778
std         348.098907
min      -14261.416667
25%           0.083333
50%           0.851528
75%           7.000000
max       15335.998056
Name: AGE, dtype: float64


### Time to Report a Crime

In [30]:
print('==============================================================')
print('-<Examine the frequency of time it takes to report a of crimes/offenses >===')
# hours
temp = df['TIME_TO_REPORT'] / 3600

print(temp.describe())
print ('-----------------------')

# excluding temp two std away from mean
print (temp[~(np.abs(temp - temp.mean())>(2*temp.std()))].describe())
print('==============================================================')

-<Examine the frequency of time it takes to report a of crimes/offenses >===
count    231500.000000
mean         52.911765
std        3900.940836
min     -788927.866667
25%           0.100000
50%           0.916667
75%           3.033333
max      876936.004722
Name: TIME_TO_REPORT, dtype: float64
-----------------------
count    231333.000000
mean         25.469048
std         198.946217
min       -7004.500000
25%           0.100000
50%           0.916667
75%           3.018889
max        7810.550000
Name: TIME_TO_REPORT, dtype: float64


In [31]:
df.to_csv('./input/dc_crime2.csv')